paasta_tools/api/views/instance.py

```diff
@@ -0,0 +1,133 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+PaaSTA service instance status/start/stop etc.
+"""
+import json
+import logging
+import traceback
+
+from pyramid.response import Response
+from pyramid.view import view_config
+
+from paasta_tools import marathon_tools
+from paasta_tools.api import settings
+from paasta_tools.cli.cmds.status import get_actual_deployments
+from paasta_tools.marathon_serviceinit import deploy_status_marathon
+from paasta_tools.marathon_serviceinit import MarathonDeployStatus
+from paasta_tools.paasta_serviceinit import get_deployment_version
+from paasta_tools.utils import NoDockerImageError
+from paasta_tools.utils import validate_service_instance
+
+
+log = logging.getLogger(__name__)
+
+
+def chronos_instance_status(instance_status, service, instance, verbose):
+    return
+
+
+def marathon_job_status(client, job_config):
+    mstatus = {}
+
+    try:
+        app_id = job_config.format_marathon_app_dict()['id']
+        mstatus['app_id'] = app_id
+    except NoDockerImageError:
+        error_msg = "Docker image is not in deployments.json."
+        mstatus['message'] = error_msg
+        return mstatus
+
+    if marathon_tools.is_app_id_running(app_id, client):
+        app = client.get_app(app_id)
+        deploy_status, _ = deploy_status_marathon(app, app_id, client)
+        mstatus['deploy_status'] = MarathonDeployStatus.tostring(deploy_status)
+
+        # by comparing running count with expected count, callers can figure
+        # out if the instance is in Healthy, Warning or Critical state.
+        mstatus['running_instance_count'] = app.tasks_running
+        mstatus['expected_instance_count'] = job_config.get_instances()
+    else:
+        mstatus['deploy_status'] = 'Not Running'
+
+    return mstatus
+
+
+def marathon_instance_status(instance_status, service, instance, verbose):
+    marathon_config = marathon_tools.load_marathon_config()
+    client = marathon_tools.get_marathon_client(marathon_config.get_url(),
+                                                marathon_config.get_username(),
+                                                marathon_config.get_password())
+
+    apps = marathon_tools.get_matching_appids(service, instance, client)
+    job_config = marathon_tools.load_marathon_service_config(
+        service, instance, settings.cluster, soa_dir=settings.soa_dir)
+
+    # bouncing status can be inferred from app_count, ref get_bouncing_status
+    instance_status['app_count'] = len(apps)
+    instance_status['bounce_method'] = job_config.get_bounce_method()
+    instance_status['desired_state'] = job_config.get_desired_state()
+
+    instance_status['marathon'] = marathon_job_status(client, job_config)
+
+
+def error_response(error_message, status_code):
```

Comment:
The common Pyramid pattern to handle this is to make your own `Exception` class, raise it and provide a handler for it that does this logic.

See: http://docs.pylonsproject.org/projects/pyramid_cookbook/en/latest/pylons/exceptions.html#exception-views